python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py

```diff
@@ -136,100 +142,345 @@ def convert_to_static(function):
+        self.input_with_spec = input_with_spec
+        self.class_instance = class_instance
+
+    @classmethod
+    def from_func_and_args(cls, function_spec, args, kwargs, class_instance):
         """
-        params = collections.OrderedDict()
-        if self.is_method():
-            layer_instance = self._args[0]
-            if include_sublayer:
-                params = layer_instance.parameters()
-                names = [p.name for p in params]
-                params = collections.OrderedDict(zip(names, params))
+        Generated a CacheKey instance by given inputs.
+
+        Args:
+            functions_spec(FunctionSpec): a FunctionSpec instance of decorated function.
+            args(tuple): tuple of actual inputs arguments.
+            kwargs(dict): dict of actual inputs keyword arguments.
+            class_instance(object): a instance of class `Layer`.
+        """
+        # 1. filter `self` in args
+        if args and isinstance(args[0], layers.Layer):
+            args = args[1:]
+        # 2. convert tensor and numpy array into InputSpec 
+        _args, _kwargs = function_spec.unified_args_and_kwargs(args, kwargs)
+        input_with_spec = function_spec.args_to_input_spec(_args, _kwargs)
+
+        # 3. check whether hit the cache or build a new program for the input arguments
+        return CacheKey(function_spec, input_with_spec, class_instance)
+
+    def __hash__(self):
+        return hash((id(self.function_spec),
+                     _make_hashable(self.input_with_spec), self.class_instance))
+
+    def __eq__(self, other):
+        return (type(self) is type(other)) and hash(self) == hash(other)
+
+    def __neq__(self, other):
+        return not self == other
+
+    def __repr__(self):
+        return "id(function_spec): {}, input_with_spec: {}, class_instance: {}".format(
+            id(self.function_spec), self.input_with_spec, self.class_instance)
+
+
+def _make_hashable(x):
+    """
+    Makes input `x` hashable.
+
+    `inputs_with_spec` may contain some unhashable object, such as `dict/list/np.ndarray`.
+    Apply hash function using their values.
+    """
+    if isinstance(x, (tuple, list)):
+        return tuple(map(_make_hashable, x))
+
+    try:
+        hash(x)
+    except TypeError:
+        if isinstance(x, np.ndarray):
+            # Note: `tostring()` will return the binary data from np.ndarray that
+            # means different value will lead to different hash code.
+            return hash(x.tostring())
+        elif isinstance(x, dict):
+            return tuple(map(_make_hashable, x.values()))
+
+        raise ValueError("Arguments to a `@declarative` must be a hashable"
+                         "Python objects (or nested structures of these types)."
+                         " But received type: %s" % type_name(x))
+
+    return x
+
+
+def unwrap_decorators(func):
+    """
+    Unwraps a decorated function and returns the decorator list and inner target.
```

Comment:
Is this for function or for StaticLayer? I saw the code is StaticLayer. If it is used for function, Yamei has implemented in the past, can you check whether that can be reused?

Reply:
This is a different function.